relayer/chains/substrate/construct.go

```diff
@@ -0,0 +1,682 @@
+			return sortedParaIds[i] < sortedParaIds[j]
+		})
+
+		for _, paraId := range sortedParaIds {
+			paraIdScale := make([]byte, 4)
+			// scale encode para_id
+			binary.LittleEndian.PutUint32(paraIdScale[:], paraId)
+			leaf := append(paraIdScale, paraHeaders[paraId]...)
+			paraHeadsLeaves = append(paraHeadsLeaves, crypto.Keccak256(leaf))
+			if paraId == sp.Config.ParaID {
+				// note index of paraId
+				index = uint64(count)
+			}
+			count++
+		}
+
+		tree, err := merkle.NewTree(hasher.Keccak256Hasher{}).FromLeaves(paraHeadsLeaves)
+		if err != nil {
+			return nil, err
+		}
+		paraHeadsProof := tree.Proof([]uint64{index})
+		authorityRoot := bytes32(v.BeefyNextAuthoritySet.Root[:])
+		parentHash := bytes32(v.ParentNumberAndHash.Hash[:])
+
+		parachainHeaderDecoded, err := beefyclienttypes.DecodeParachainHeader(paraHeaders[sp.Config.ParaID])
+		if err != nil {
+			return nil, err
+		}
+
+		timestampExt, extProof, err := sp.constructExtrinsics(uint32(parachainHeaderDecoded.Number))
+		if err != nil {
+			return nil, err
+		}
+
+		header := beefyclienttypes.ParachainHeader{
+			ParachainHeader: paraHeaders[sp.Config.ParaID],
+			PartialMMRLeaf: &beefyclienttypes.PartialMMRLeaf{
+				Version:      beefyclienttypes.U8(v.Version),
+				ParentNumber: uint32(v.ParentNumberAndHash.ParentNumber),
+				ParentHash:   &parentHash,
+				BeefyNextAuthoritySet: beefyclienttypes.BeefyAuthoritySet{
+					ID:            uint64(v.BeefyNextAuthoritySet.ID),
+					Len:           uint32(v.BeefyNextAuthoritySet.Len),
+					AuthorityRoot: &authorityRoot,
+				},
+			},
+			ParachainHeadsProof: paraHeadsProof.ProofHashes(),
+			HeadsLeafIndex:      uint32(index),
+			HeadsTotalCount:     uint32(len(paraHeadsLeaves)),
+			TimestampExtrinsic:  timestampExt,
+			ExtrinsicProof:      extProof,
+		}
+
+		parachainHeaders = append(parachainHeaders, &header)
+	}
+
+	return parachainHeaders, nil
+}
+
+func (sp *SubstrateProvider) constructExtrinsics(
+	blockNumber uint32,
+) (timestampExtrinsic []byte, extrinsicProof [][]byte, err error) {
+	blockHash, err := sp.RPCClient.RPC.Chain.GetBlockHash(uint64(blockNumber))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	block, err := sp.RPCClient.RPC.Chain.GetBlock(blockHash)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	exts := block.Block.Extrinsics
+	if len(exts) == 0 {
+		return nil, nil, nil
+	}
+
+	timestampExtrinsic, err = Encode(exts[0])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	t := trie.NewEmptyTrie()
+	for i := 0; i < len(exts); i++ {
+		ext, err := Encode(exts[i])
+		if err != nil {
+			return nil, nil, err
+		}
+
+		key := rpcclienttypes.NewUCompactFromUInt(uint64(i))
+		encodedKey, err := Encode(key)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		t.Put(encodedKey, ext)
+	}
+
+	userHomeDir, err := os.UserHomeDir()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	memdb, err := chaindb.NewBadgerDB(&chaindb.Config{
+		InMemory: true,
+		DataDir:  userHomeDir,
```

Comment:
I think ideally this should be the same directory where the relayer keys and config are located (which the default location is `~/.relayer/`). Leaking files related to the relayer outside of the home directory where other relayer files are written to disk is confusing for users and can quickly create a mess of the filesystem

The home directory value is passed into the `NewProvider` call so maybe it makes sense to store this value on the `SubstrateProvider` so it can be accessed here later on.